### PR TITLE
[Tidy] Store `page.id` in outer page container to enable page specific styling

### PR DESCRIPTION
**Pull request by huong-li-nguyen**
## Description

- Store `page.id` in outer page container to enable page specific stylings (e.g. this now enables hiding the page-title selectively on a given page)
- Change `page-container` to be a className instead of an ID
- Add quick CSS fix that aligns table and graph titles when next to each other (see dev example)
Screenshot 2024-05-06 at 14 07 02

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
  - I have not referenced individuals, products or companies in any commits, directly or indirectly.
  - I have not added data or restricted code in any commits, directly or indirectly.

### CODE REVIEW

File: vizro-core/src/vizro/models/_models_utils.py
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/models/_models_utils.py:27]: This function had to be refactored for the sake of creating valid CSS IDs e.g. a valid CSS ID cannot start with` / `

File: vizro-core/tests/unit/vizro/models/test_models_utils.py
Comment (yury-fedotov) [inline vizro-core/tests/unit/vizro/models/test_models_utils.py:58]: Was opening an issue just now and randomly came across this :) You can do multiple `parametrize` on the same test, which is like a nested for loop, so those two tests can be joined into one if you like.

```suggestion
@pytest.mark.parametrize("enable_url", [True, False])
@pytest.mark.parametrize(
    "test_string, expected",
    [
        ("Title", "title"),
        ("Page Title", "page-title"),
        ("Page Title!", "page-title"),
        ("Another123Title", "another123title"),
        ("Some_Example%Title", "some_exampletitle"),
        ("!@#$%^&*()_+=", "_"),
        ("", ""),
    ],
)
def test_clean_path_with_url_enabled(test_string, expected, enable_url):
    cleaned_string = _clean_path(path=test_string, allowed_characters="-_/", enable_url=enable_url)
    expected_output = f"/{expected}" if enable_url else expected
    assert cleaned_string == expected_output
```

Comment (huong-li-nguyen) [reply vizro-core/tests/unit/vizro/models/test_models_utils.py:58]: Haha, amazing! 🚀 I actually parameterised all four arguments before the way you proposed it above, but I thought it wasn't easy to read. I like your suggestion above though, as it is still kept very simple! Thanks!!

Comment (yury-fedotov) [reply vizro-core/tests/unit/vizro/models/test_models_utils.py]: Welcome. Also might be helpful to have type hints in test functions signatures too (that would show that last param is `bool` while two others are `str`, which is probably what you meant by "difficult to read").

But I didn't add it in my example to keep it isolated to the double parametrization only.

Comment (huong-li-nguyen) [reply vizro-core/tests/unit/vizro/models/test_models_utils.py]: That's a great suggestion 👍 We do the type hints on the actual functions we test so we have it there, but I don't think we do it on the test function signatures yet. 

I had this before and then a long list of use cases 😄 In the end I thought, it became more difficult to read and understand because in the end `enable_url=True` just adds a `/` at the beginning, but that didn't become immediately clear from screening through all the use cases. Didn't think about just adding a simple if/else - haha! Thanks for the suggestion and your first contribution! 🚀 

```
@pytest.mark.parametrize(
    "test_string, allowed_characters, enable_url, expected",
    [
        ("Title", "-_/", True, "/title"),
        ("Title", "-_/", False, "title"),
        ...
    ],
)
def test_clean_path_with_url_enabled(test_string, allowed_characters, enable_url, expected):
    cleaned_string = _clean_path(path=test_string, allowed_characters=allowed_characters, enable_url=enable_url)
    assert cleaned_string == expected_output
```

**Review by nadijagraca — APPROVED** [review]
LGTM. ⭐

File: vizro-core/docs/pages/user-guides/assets.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/assets.md:156]: Is this section still correct? e.g. it talks a lot about `_outer`, and I'm not sure whether that's still used much after your recent bootstrap refactorings?

File: vizro-core/docs/pages/user-guides/assets.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/assets.md:93]:
```suggestion
For instance, if a `page.id` is `Variable Analysis` then it has CSS ID `variable-analysis`.
```

As it stands the "," and "." are very confusing here because it looks like you're replacing "," with ".".

Not sure whether we should convert to lowercase, but let's discuss 🤔

Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/assets.md:93]: Oh, I don't even know how this got in - probably my automatic grammar checker😆 There is not supposed to be a comma 😄

File: vizro-core/docs/pages/user-guides/assets.md
Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/assets.md:156]: This still holds, but it gets a bit confusing now because it does not hold for all components. Let's discuss in our call, but just to show you what I mean:

Here we don't have the `_outer` suffix for an ID, because we can provide the self.id directly to the component that gets outputted:

`dbc.Button(id=self.id, children=self.text)`

However, here we still need the `_outer` suffix because of two reasons:

1. We once aligned that the `self.id` is being provided to the inner component that might be reactive to callbacks. In the example of the card it is the `dcc.Markdown` component. Because of that we cannot provide the `self.id` to the outermost component. 
2. Let's say someone wants to change the card background-color but only for a selective card and not for all. Here we need to provide an `id` to the outermost wrapper of the card so we can selectively target all inner components. And because we can't use `self.id`, we started using `self.id_outer`.

Happy to change this current system though, just not sure what's the best approach here 🤔 

`dbc.Card(card_content, className=card_class, id=f"{self.id}_outer")`

Comment (antonymilne) [reply vizro-core/docs/pages/user-guides/assets.md:156]: Yeah, that is what I expected but I didn't know how prevalent the `_outer` suffix was any more.

I think the current system basically makes sense (using `self.id` for the "core" reactive component rather than the outer container) but we should _consider_ (but maybe not do...) removing the `_outer` ones if they're just there for CSS selectors [in favour of using `has()`](https://stackoverflow.com/questions/1014861/is-there-a-css-parent-selector).

File: vizro-core/src/vizro/models/_dashboard.py
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/models/_dashboard.py]: Ah yes! Implemented here: https://github.com/mckinsey/vizro/pull/455/commits/ba9966e050ad1070f9e99dd3f0c0a50da6ded65f

File: vizro-core/docs/pages/user-guides/assets.md
Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/assets.md:156]: Yes, I like that! Let's discuss in our call later. But I would also be in favour of killing the entire _outer system and just updating the docs 👍

File: vizro-core/src/vizro/models/_models_utils.py
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/models/_models_utils.py:27]: Just for my own tracking: We've decided to **not clean** the path, but instead add to the docs that the user has to provide a valid `id`. I'll add to the docs what we consider as `valid`:
- Use lowercase letters and hyphens
- Avoid using special characters or spaces
- Provide a unique `id`

File: vizro-core/docs/pages/user-guides/assets.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/assets.md:97]:
```suggestion
To style components for only a specific page, use the page's `id` in CSS selectors. By default, this is the [same as the page `title`](pages.md), but such a value might not be a valid CSS identifier. A suitable `id` must be unique across all models in the dashboard and should contain only alphanumeric characters, hyphens (`-`) and underscores (`_`). In particular, note that spaces are _not_ allowed.
```

This guidance is general about `id` really applies everywhere of course, not just to page's. Feel free to move it now or in the future when we restructure the page as discussed earlier.

The only special part that applies specifically to page is just that the `id` is automatically set and hence more likely to be inappropriate for CSS identifiers due to having a space in it.

File: vizro-core/docs/pages/user-guides/assets.md
Comment (antonymilne) [inline vizro-core/docs/pages/user-guides/assets.md:103]:
```suggestion
1. Provide a valid `id` to the `Page`, for example `Page(id="page-with-hidden-title", title="Page with hidden title", ...)`.
2. Identify the CSS class or ID you need to target. To hide the page title, you need to hide the container with the ID `right-header`.
3. Use the `id` in combination with CSS selectors to change the relevant CSS properties.
4. Add your custom css file to the `assets` folder as explained above.
```

**Review by antonymilne — APPROVED** [review]
Thank you for the changes, looks good!

**Comment by huong-li-nguyen** [comment]
@stichbury - do you want to take a final look at the docs changes? :) I'll merge after your approval 👍

File: vizro-core/docs/pages/user-guides/assets.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/assets.md:91]:
```suggestion
To style components for a specific page, use the page's `id` in CSS selectors. By default, this is the [same as the page `title`](pages.md), but such a value might not be a valid CSS identifier. A suitable `id` must be unique across all models in the dashboard and should contain only alphanumeric characters, hyphens (`-`) and underscores (`_`). In particular, note that spaces are _not_ allowed.
```

**Review by stichbury — APPROVED** [review]
LGTM, thank you 🙏

**Pull request merged**
